File: app/infrastructure/jira/jira_rest_client.py
"""Adapter concreto de JiraClient contra a Jira Cloud REST API v3.

Contrato verificado em 2026-08-24 contra a documentação oficial da Atlassian:
- Endpoint legado /rest/api/3/search foi REMOVIDO. Usamos /rest/api/3/search/jql.
- Paginação por `nextPageToken` (não mais `startAt`) — token expira em 7 dias, não
  cacheamos entre execuções.
- Autenticação: Basic Auth com email + API token (não senha).
- `fields.subtasks` retorna só id/key/summary/issuetype por padrão; para pegar o
  corpo de texto da subtarefa Release Note é preciso uma segunda chamada em
  /rest/api/3/issue/{key} pedindo o campo `description`.

Fontes: developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-issue-search,
confluence.atlassian.com/jirakb/run-jql-search-query-using-jira-cloud-rest-api.
"""

from __future__ import annotations

import httpx
from loguru import logger
from tenacity import retry, retry_if_exception_type, stop_after_attempt, wait_exponential

from app.domain.entities import HistoriaJira

_TIPO_ISSUE_RELEASE_NOTE = "Release Note"  # nome do subtask type no Jira da InvoiSys —
# confirmar exato quando tivermos acesso à instância real (pendência rastreada
# na documentação interna de estado do projeto)


class JiraApiError(Exception):
    """Erro de comunicação com a API do Jira, após esgotar as tentativas de retry."""


class JiraRestClient:
    """Implementa a porta `app.domain.ports.JiraClient`."""

    def __init__(
        self,
        base_url: str,
        email: str,
        api_token: str,
        http_client: httpx.AsyncClient | None = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._auth = (email, api_token)
        # Cliente injetável — em teste, passamos um AsyncClient com transport mockado
        # em vez de bater na rede de verdade.
        self._http = http_client or httpx.AsyncClient(
            base_url=self._base_url,
            auth=self._auth,
            headers={"Accept": "application/json"},
            timeout=httpx.Timeout(15.0, connect=5.0),
        )

    async def buscar_historias_da_release(self, fix_version: str) -> list[HistoriaJira]:
        issues_raw = await self._buscar_todas_issues(fix_version)
        historias: list[HistoriaJira] = []
        for issue in issues_raw:
            texto_release_note = await self._buscar_texto_release_note(issue)
            historias.append(self._to_historia_jira(issue, texto_release_note))
        return historias

    async def _buscar_todas_issues(self, fix_version: str) -> list[dict]:
        jql = f'fixVersion = "{fix_version}"'
        issues: list[dict] = []
        next_page_token: str | None = None

        while True:
            params: dict[str, str | int] = {
                "jql": jql,
                "fields": "summary,description,issuetype,labels,subtasks",
                "maxResults": 100,
            }
            if next_page_token:
                params["nextPageToken"] = next_page_token

            payload = await self._get("/rest/api/3/search/jql", params)
            issues.extend(payload.get("issues", []))

            next_page_token = payload.get("nextPageToken")
            if not next_page_token or payload.get("isLast", True):
                break

        return issues

    async def _buscar_texto_release_note(self, issue: dict) -> str | None:
        """Procura, entre as subtarefas da issue, uma do tipo Release Note e retorna
        seu texto. Custa uma chamada extra por subtarefa candidata — aceitável no
        volume de uma Release (dezenas de issues, não milhares)."""
        subtasks = issue.get("fields", {}).get("subtasks", [])
        for subtask in subtasks:
            tipo = subtask.get("fields", {}).get("issuetype", {}).get("name", "")
            if tipo == _TIPO_ISSUE_RELEASE_NOTE:
                detalhe = await self._get(
                    f"/rest/api/3/issue/{subtask['key']}",
                    {"fields": "description"},
                )
                return self._extrair_texto_description(detalhe.get("fields", {}))
        return None

    @staticmethod
    def _to_historia_jira(issue: dict, texto_release_note: str | None) -> HistoriaJira:
        fields = issue.get("fields", {})
        return HistoriaJira(
            chave=issue["key"],
            titulo=fields.get("summary", ""),
            descricao_tecnica=JiraRestClient._extrair_texto_description(fields),
            tipo_issue=fields.get("issuetype", {}).get("name", ""),
            texto_release_note=texto_release_note,
            labels=fields.get("labels", []),
        )

    @staticmethod
    def _extrair_texto_description(fields: dict) -> str:
        """O campo `description` da API v3 vem em Atlassian Document Format (ADF),
        não texto puro. Extração completa de ADF (tabelas, listas, formatação) fica
        para quando tivermos exemplos reais do Jira da InvoiSys — por ora extrai só
        os nós de texto simples, suficiente para alimentar o pipeline de IA."""
        description = fields.get("description")
        if not description:
            return ""
        if isinstance(description, str):
            return description  # instância antiga/config diferente pode devolver string direto
        return JiraRestClient._extrair_texto_adf(description)

    @staticmethod
    def _extrair_texto_adf(node: dict) -> str:
        partes: list[str] = []
        if node.get("type") == "text":
            partes.append(node.get("text", ""))
        for filho in node.get("content", []):
            partes.append(JiraRestClient._extrair_texto_adf(filho))
        return " ".join(p for p in partes if p).strip()

    @retry(
        retry=retry_if_exception_type(httpx.TransportError),
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=1, min=1, max=8),
        reraise=True,
    )
    async def _get(self, path: str, params: dict) -> dict:
        try:
            response = await self._http.get(path, params=params)
            response.raise_for_status()
            return response.json()
        except httpx.HTTPStatusError as exc:
            logger.error(
                "Jira respondeu erro {} em {}: {}",
                exc.response.status_code,
                path,
                exc.response.text[:500],
            )
            raise JiraApiError(f"Jira retornou {exc.response.status_code} em {path}") from exc

    async def aclose(self) -> None:
        await self._http.aclose()

File: src/InvoiSys.Infrastructure/Jira/JiraRestClient.cs
using System.Net.Http.Json;
using System.Text.Json;
using System.Web;
using InvoiSys.Domain.Entities;
using InvoiSys.Domain.Ports;
using Microsoft.Extensions.Logging;

namespace InvoiSys.Infrastructure.Jira;

/// <summary>
/// Adapter concreto de <see cref="IJiraClient"/> contra a Jira Cloud REST API v3.
///
/// Contrato verificado em 2026-08-24 contra a documentação oficial da Atlassian:
/// - Endpoint legado /rest/api/3/search foi REMOVIDO. Usamos /rest/api/3/search/jql.
/// - Paginação por nextPageToken (não mais startAt) — token expira em 7 dias, não
///   cacheamos entre execuções.
/// - Autenticação: Basic Auth com email + API token (não senha) — configurada no
///   HttpClient nomeado, na composition root.
/// - fields.subtasks retorna só id/key/summary/issuetype por padrão; para pegar o
///   corpo de texto da subtarefa Release Note é preciso uma segunda chamada em
///   /rest/api/3/issue/{key} pedindo o campo description.
///
/// Retry (429/5xx e falha de transporte) fica na policy de resiliência registrada no
/// HttpClient, não aqui — o adapter só traduz HTTP em domínio.
/// </summary>
public sealed class JiraRestClient(HttpClient http, ILogger<JiraRestClient> logger) : IJiraClient
{
    /// <summary>
    /// Nome do subtask type no Jira da InvoiSys — confirmar exato quando tivermos
    /// acesso à instância real (pendência rastreada na documentação interna).
    /// </summary>
    private const string TipoIssueReleaseNote = "Release Note";

    public async Task<IReadOnlyList<HistoriaJira>> BuscarHistoriasDaReleaseAsync(
        string fixVersion,
        CancellationToken cancellationToken = default)
    {
        var issues = await BuscarTodasIssuesAsync(fixVersion, cancellationToken);

        var historias = new List<HistoriaJira>(issues.Count);
        foreach (var issue in issues)
        {
            var textoReleaseNote = await BuscarTextoReleaseNoteAsync(issue, cancellationToken);
            historias.Add(ParaHistoriaJira(issue, textoReleaseNote));
        }

        return historias;
    }

    private async Task<List<JsonElement>> BuscarTodasIssuesAsync(
        string fixVersion,
        CancellationToken cancellationToken)
    {
        var jql = $"fixVersion = \"{fixVersion}\"";
        var issues = new List<JsonElement>();
        string? nextPageToken = null;

        while (true)
        {
            var query = HttpUtility.ParseQueryString(string.Empty);
            query["jql"] = jql;
            query["fields"] = "summary,description,issuetype,labels,subtasks";
            query["maxResults"] = "100";
            if (nextPageToken is not null)
            {
                query["nextPageToken"] = nextPageToken;
            }

            var payload = await GetAsync($"/rest/api/3/search/jql?{query}", cancellationToken);

            if (payload.TryGetProperty("issues", out var lista)
                && lista.ValueKind == JsonValueKind.Array)
            {
                issues.AddRange(lista.EnumerateArray());
            }

            nextPageToken = payload.TryGetProperty("nextPageToken", out var token)
                && token.ValueKind == JsonValueKind.String
                    ? token.GetString()
                    : null;

            var ehUltima = !payload.TryGetProperty("isLast", out var isLast)
                || isLast.ValueKind != JsonValueKind.False;

            if (string.IsNullOrEmpty(nextPageToken) || ehUltima)
            {
                break;
            }
        }

        return issues;
    }

    /// <summary>
    /// Procura, entre as subtarefas da issue, uma do tipo Release Note e retorna seu
    /// texto. Custa uma chamada extra por subtarefa candidata — aceitável no volume de
    /// uma Release (dezenas de issues, não milhares).
    /// </summary>
    private async Task<string?> BuscarTextoReleaseNoteAsync(
        JsonElement issue,
        CancellationToken cancellationToken)
    {
        if (!issue.TryGetProperty("fields", out var fields)
            || !fields.TryGetProperty("subtasks", out var subtasks)
            || subtasks.ValueKind != JsonValueKind.Array)
        {
            return null;
        }

        foreach (var subtask in subtasks.EnumerateArray())
        {
            var tipo = subtask.TryGetProperty("fields", out var subFields)
                && subFields.TryGetProperty("issuetype", out var issuetype)
                && issuetype.TryGetProperty("name", out var nome)
                    ? nome.GetString()
                    : null;

            if (tipo != TipoIssueReleaseNote)
            {
                continue;
            }

            var chave = subtask.GetProperty("key").GetString();
            var detalhe = await GetAsync(
                $"/rest/api/3/issue/{chave}?fields=description",
                cancellationToken);

            return detalhe.TryGetProperty("fields", out var detalheFields)
                ? ExtrairTextoDescription(detalheFields)
                : null;
        }

        return null;
    }

    private static HistoriaJira ParaHistoriaJira(JsonElement issue, string? textoReleaseNote)
    {
        var fields = issue.TryGetProperty("fields", out var f) ? f : default;

        return new HistoriaJira
        {
            Chave = issue.GetProperty("key").GetString() ?? string.Empty,
            Titulo = LerString(fields, "summary"),
            DescricaoTecnica = ExtrairTextoDescription(fields),
            TipoIssue = fields.ValueKind == JsonValueKind.Object
                && fields.TryGetProperty("issuetype", out var tipo)
                    ? LerString(tipo, "name")
                    : string.Empty,
            TextoReleaseNote = textoReleaseNote,
            Labels = fields.ValueKind == JsonValueKind.Object
                && fields.TryGetProperty("labels", out var labels)
                && labels.ValueKind == JsonValueKind.Array
                    ? [.. labels.EnumerateArray()
                        .Select(l => l.GetString())
                        .Where(l => l is not null)
                        .Select(l => l!)]
                    : [],
        };
    }

    /// <summary>
    /// O campo description da API v3 vem em Atlassian Document Format (ADF), não texto
    /// puro. Extração completa de ADF (tabelas, listas, formatação) fica para quando
    /// tivermos exemplos reais do Jira da InvoiSys — por ora extrai só os nós de texto
    /// simples, suficiente para alimentar o pipeline de IA.
    /// </summary>
    internal static string ExtrairTextoDescription(JsonElement fields)
    {
        if (fields.ValueKind != JsonValueKind.Object
            || !fields.TryGetProperty("description", out var description))
        {
            return string.Empty;
        }

        return description.ValueKind switch
        {
            // Instância antiga/config diferente pode devolver string direto.
            JsonValueKind.String => description.GetString() ?? string.Empty,
            JsonValueKind.Object => ExtrairTextoAdf(description),
            _ => string.Empty,
        };
    }

    internal static string ExtrairTextoAdf(JsonElement node)
    {
        var partes = new List<string>();
        ColetarTextoAdf(node, partes);
        return string.Join(' ', partes.Where(p => !string.IsNullOrWhiteSpace(p))).Trim();
    }

    private static void ColetarTextoAdf(JsonElement node, List<string> partes)
    {
        if (node.ValueKind != JsonValueKind.Object)
        {
            return;
        }

        if (node.TryGetProperty("type", out var tipo)
            && tipo.ValueKind == JsonValueKind.String
            && tipo.GetString() == "text"
            && node.TryGetProperty("text", out var texto))
        {
            partes.Add(texto.GetString() ?? string.Empty);
        }

        if (node.TryGetProperty("content", out var content)
            && content.ValueKind == JsonValueKind.Array)
        {
            foreach (var filho in content.EnumerateArray())
            {
                ColetarTextoAdf(filho, partes);
            }
        }
    }

    private static string LerString(JsonElement elemento, string propriedade) =>
        elemento.ValueKind == JsonValueKind.Object
        && elemento.TryGetProperty(propriedade, out var valor)
        && valor.ValueKind == JsonValueKind.String
            ? valor.GetString() ?? string.Empty
            : string.Empty;

    private async Task<JsonElement> GetAsync(string caminho, CancellationToken cancellationToken)
    {
        using var resposta = await http.GetAsync(caminho, cancellationToken);

        if (!resposta.IsSuccessStatusCode)
        {
            var corpo = await resposta.Content.ReadAsStringAsync(cancellationToken);
            corpo = corpo.Length > 500 ? corpo[..500] : corpo;

            logger.LogError(
                "Jira respondeu erro {Status} em {Caminho}: {Corpo}",
                (int)resposta.StatusCode,
                caminho,
                corpo);

            // O corpo fica só no log acima: pode trazer detalhe da conta/instância, e esta
            // mensagem chega até quem chamou a API.
            throw new JiraApiException(
                $"Jira retornou {(int)resposta.StatusCode} em {caminho}.");
        }

        return await resposta.Content.ReadFromJsonAsync<JsonElement>(cancellationToken);
    }
}
